### PR TITLE
[cmake] Provide private extension points to the build

### PR DESCRIPTION
**Pull request by etcwilde**
This patch provides extension points on the build system so that vendors can keep vendor-specific build system settings private, without having to worry about merge conflicts haunting them for the rest of time.

The location of the extension modules is set with the `SwiftCore_PRIVATE_MODULE_DIR` variable. If it is not specified, the default location is under `cmake/modules/private`.

Currently, there are extension points for the default settings, and for the global settings.

### CODE REVIEW

**Comment by etcwilde** [comment]
@swift-ci please test

File: Runtimes/Core/CMakeLists.txt [hunk 1]
```diff
@@ -47,6 +48,13 @@ set(SwiftCore_SWIFTC_SOURCE_DIR
   "${PROJECT_SOURCE_DIR}/../../"
   CACHE FILEPATH "Path to the root source directory of the Swift compiler")
 
+# Hook point for vendor-specific private extensions to the build system
+# Allowed extension points:
+#   - DefaultSettings.cmake
+#   - Settings.cmake
+set(SwiftCore_PRIVATE_MODULE_DIR "${SwiftCore_CMAKE_MODULES_DIR}/private"
```
Comment (compnerd) [inline Runtimes/Core/CMakeLists.txt:55]: What do you think of `SwiftCore_VENDOR_MODULE_DIR`? This is the vendor specific settings and makes it clear when you concatenate them even if you don't know what it is (e.g. `${SwiftCore_VENDOR_MODULE_DIR}/DefaultSettings.cmake` vs `${SwiftCore_PRIVATE_MODULE_DIR}/DefaultSettings.cmake`).

**Review by compnerd — APPROVED** [review]

File: Runtimes/Core/CMakeLists.txt (hunk 1 above)
Comment (etcwilde) [reply Runtimes/Core/CMakeLists.txt:55]: Sure, sounds good to me

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Review by mikeash — APPROVED** [review]
Excellent!

**Comment by etcwilde** [comment]
@swift-ci please smoke test Linux

**Pull request merged**